miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultSampleClassService.java

```diff
@@ -78,4 +85,25 @@ public void delete(Long sampleClassId) throws IOException {
     sampleClassDao.deleteSampleClass(sampleClass);
   }
 
+  @Override
+  public SampleClass inferStockFromAliquot(SampleClass sampleClass) {
+    SampleClass aliquotClass = sampleClassDao.getSampleClass(sampleClass.getId());
+    if (aliquotClass == null || !aliquotClass.getSampleCategory().equals(SampleAliquot.CATEGORY_NAME)) {
+      throw new IllegalArgumentException("Sample class " + sampleClass.getId() + " is not a valid aliquot class.");
+    }
+    SampleClass stockClass = null;
+    for (SampleValidRelationship relationship : sampleValidRelationshipDao.getSampleValidRelationship()) {
+      if (relationship.getChild().getId() == aliquotClass.getId()
+          && relationship.getParent().getSampleCategory().equals(SampleStock.CATEGORY_NAME)) {
```

Comment:
Make sure it's also not archived. cDNA (aliquot) can be parented to cDNA (stock) or whole RNA (stock), the latter of which is an archived relationship